core/mutator/entry/mutation.go

```diff
@@ -0,0 +1,134 @@
+// Copyright 2017 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package entry
+
+import (
+	"github.com/benlaurie/objecthash/go/objecthash"
+	"github.com/golang/protobuf/proto"
+	"github.com/google/keytransparency/core/crypto/commitments"
+	"github.com/google/keytransparency/core/crypto/signatures"
+	"github.com/google/keytransparency/core/mutator"
+	tpb "github.com/google/keytransparency/core/proto/keytransparency_v1_types"
+	"github.com/google/trillian/crypto/sigpb"
+)
+
+// Mutation provides APIs for manipulating entries.
+type Mutation struct {
+	userID, appID string
+	index         []byte
+	data, nonce   []byte
+
+	prevEntry *tpb.Entry
+	entry     *tpb.Entry
+}
+
+// NewMutation creates a mutation object from a previous value which can be modified.
+func NewMutation(oldValue, index []byte, userID, appID string) (*Mutation, error) {
+	prevEntry, err := FromLeafValue(oldValue)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO(ismail): Change this to plain sha256.
+	hash := objecthash.ObjectHash(prevEntry)
+	return &Mutation{
+		userID:    userID,
+		appID:     appID,
+		index:     index,
+		prevEntry: prevEntry,
+		entry: &tpb.Entry{
+			AuthorizedKeys: prevEntry.GetAuthorizedKeys(),
+			Previous:       hash[:],
+			Commitment:     prevEntry.GetCommitment(),
+		},
+	}, nil
+}
+
+// SetCommitment updates entry to be a commitment to data.
+func (m *Mutation) SetCommitment(data []byte) error {
```

Comment:
Can we combine this function with `NewMutation`? `NewMutation` can take data and return a new mutation object with the commitment set.